### PR TITLE
rgw: orphan fixes 

### PR DESCRIPTION
**Pull request by theanalyst**
I saw an issue in luminous where orphans find was running really slowly. Reproducing on my vstart cluster, with a single bucket having 2 million+ entries, I was able to see orphans find taking upwards of 24 hours in a vstart cluster on SSDs. The search was continually in the third bucket index stage. Stopping the IO resulted in orphans find still taking around 6.5 hours to complete.

Upon analyzing further I saw that we ignore head objects in the initial rados listing, however we do account for, stat and store the entry in the linked entry. I  added a check after the rados stat stage where we check the size of OLH and compare against the size of the object. If the object size fits within the manifest then we can safely avoid logging this object as it would have been skipped anyway.  I further assume clusters from luminous with object sizes < 4MiB would not have a shadow object and we could reduce a rados stat as well as an omap entry which we would ignore in the end as head objects are dropped in the initial rados list. For older clusters from pre Jewel era, the old behaviour can be turned back on by adding a `--detail` flag to orphans search

I also saw that we request a max_entries of 100 keys for bucket listing, however the underlying bucket still requests `rgw_list_bucket_min_readahead` entries anyway in the cls which lead to wasteful object listing as we were requesting 100 entries though the underlying layer anyway requested a 1000 entries and threw away the rest. Also workloads which continously append objects in an increasing alphabetic order would mean this stage would almost never complete as at the pace of 100 objects we may not be able to catch up with the bucket which might be able to add more in that time frame. Keeping listing in sync with the underlying cls layer helped massively speed up to less than an hours time.

I further noticed that we process all bucket instances, which meant in a bucket with million objects we would have processed every stale instance left behind as well (unless the cluster is 12.2.11+ which has all the resharding fixes). This made the cluster actually list and process the bucket many many times, so avoiding this and only comparing the actual bucket id greatly reduced the IO and I was able to bring down the total processing time to a couple of minutes. 

Another improvement which I feel could be done which is still to be implemented would be to ask for an unordered listing for buckets as we hash the object names and store them on the linked shard, and assuming ultimately that omap write actually orders the keys, then there isn't a need to request ordered listing from the bucket. 

*edit*- dropped TODO

### CODE REVIEW

**Comment by theanalyst** [comment]
I'm still testing the behaviour on master, but luminous saw a massive improvement so I assume we could get the same benefit in master as well. I'll report back once I have some numbers

**Comment by mattbenjamin** [comment]
@theanalyst thanks sincerely for this work!

File: src/rgw/rgw_orphan.cc [hunk 1]
```diff
@@ -191,6 +191,10 @@ int RGWOrphanSearch::init(const string& job_name, RGWOrphanSearchInfo *info) {
     return r;
   }
 
+  constexpr auto MAX_LIST_OBJS_ENTRIES = 100;
```
Comment (mattbenjamin) [inline src/rgw/rgw_orphan.cc]: yeah, excellent find

File: src/rgw/rgw_orphan.cc [hunk 2]
```diff
@@ -544,13 +543,12 @@ int RGWOrphanSearch::build_linked_oids_for_bucket(const string& bucket_instance_
           }
         }
       }
-      if (++count >= COUNT_BEFORE_FLUSH) {
+      if (oids.size() >= COUNT_BEFORE_FLUSH) {
```
Comment (mattbenjamin) [inline src/rgw/rgw_orphan.cc]: another brilliant find

File: src/rgw/rgw_orphan.h [hunk 3]
```diff
@@ -193,7 +195,7 @@ class RGWOrphanSearch {
     return orphan_store.write_job(search_info.job_name, state);
   }
 
-  int init(const string& job_name, RGWOrphanSearchInfo *info);
+  int init(const string& job_name, RGWOrphanSearchInfo *info, bool detailed_mode=false);
```
Comment (mattbenjamin) [inline src/rgw/rgw_orphan.h]: woot, ok

File: src/rgw/rgw_orphan.cc [hunk 4]
```diff
@@ -485,11 +486,47 @@ int RGWOrphanSearch::pop_and_handle_stat_op(map<int, list<string> >& oids, std::
...
+  if (ret < 0) {
+    if (ret == -ENOENT) {
+      /* probably raced with bucket removal */
+      return 0;
+    }
+    lderr(store->ctx()) << __func__ << ": ERROR: RGWRados::get_bucket_instance_info() returned ret=" << ret << dendl;
+    return ret;
+  }
+
+  if (cur_bucket_info.bucket.bucket_id != orphan_bucket.bucket_id) {
```
Comment (mattbenjamin) [inline src/rgw/rgw_orphan.cc]: another excellent find

**Comment by theanalyst** [comment]
updated, the test had a failure because help.t requires whitespace on newlines which my editor killed. Should be passing with the update

**Comment by liewegas** [comment]
@mattbenjamin ok to merge?  

this targets master now, so it'll need a nautilus backport.  there's no tracker ticket open so someone needs to remember to do it manually...

**Comment by theanalyst** [comment]
rerun: http://pulpito.ceph.com/abhi-2019-03-15_08:00:15-rgw-wip-abhi-testing-2019-03-14-2228-distro-basic-smithi/ looks clean

**Comment by theanalyst** [comment]
This would require a mimic/luminous backport as well, I'll create a tracker

File: src/rgw/rgw_orphan.cc [hunk 5]
```diff
@@ -185,12 +186,19 @@ int RGWOrphanStore::read_entries(const string& oid, const string& marker, map<st
...
   if (r < 0) {
     return r;
   }
 
+  constexpr int64_t MAX_LIST_OBJS_ENTRIES=100;
+
+  max_list_bucket_entries = std::max(store->ctx()->_conf->rgw_list_bucket_min_readahead,
+                                     MAX_LIST_OBJS_ENTRIES);
+
+  detailed_mode = detailed_mode;
```
Comment (ivancich) [inline src/rgw/rgw_orphan.cc]: Is the parameter hiding the member variable given it has the same name? Maybe the parameter should be `_detailed_mode`?

Comment (theanalyst) [reply src/rgw/rgw_orphan.cc]: good point, changing.

**Comment by theanalyst** [comment]
@mattbenjamin @ivancich can this go in?

**Comment by ivancich** [comment]
@theanalyst Has it gone through a QA run since you addressed the variable hiding issue? I think it would need to since the logic would have changed slightly.

Would you like me to put it through a QA run?

**Comment by ivancich** [comment]
jenkins test make check arm64

**Comment by theanalyst** [comment]
@ivancich yeah that'd be great if you could include in one of your qa runs again. The QA run was before teh var. hiding fix, so would make sense having one mroe

**Comment by ivancich** [comment]
@theanalyst I just merged this. Please add a tracker and connect things up if you'd like this backported.

**Comment by theanalyst** [comment]
created http://tracker.ceph.com/issues/39117 for backports tracking

**Pull request merged**